### PR TITLE
Tzachi libre all house coverage.

### PR DESCRIPTION
**Pull request by tzachi-dar**
This pr does the following:
1) Allow more than one phone to be used as a collector in a whole house network for libre (miaomiao/blucon/bubble). This is based on (almost) zero setup on xdrip normal channels.

2) Allow xDrip followers to see libre trend and bridge battery level.

3) Stop the sensor in a case a new sensor is detected also for bubble/direct scan/and libre wifi.

To add another receiver one has to do the following:
On the new phone:
1) Use source wizard to connect to libre. (until the new sensor has been scanned).
2) Use qr code to get security settings from current master.
3) Set the phone to master.
4) One should not start a new sensor on the new phone.

There is one improvement that I'm thinking of doing:
If the user will start a new sensor, and the user has also a follower phone on the network, the follower phone will receive BG messages from both phones.
To prevent this I'm thinking of adding a new button that will say something like: This phone is a libre receiver only. If this is set, the phone will only send messages of libreblock, and snooze alerts but not other messages (such as calibrations and BG).
Does this improvement looks good? Any suggestions for a better method?

### CODE REVIEW

**Comment by jamorham** [comment]
In principle I think this could be good but the implementation needs to be carefully handled. 

I haven't fully examined the PR but the most obvious concern is that it isn't controlled by a preference switch. So for example in a non-whole-house situation with master/follower you are still relaying the LibreBlock data which will create an overhead. In order to avoid changing the existing default behavior I think this needs to be controlled by a preference switch that defaults to off.

File: app/src/main/java/com/eveningoutpost/dexdrip/GcmActivity.java [hunk 1]
```diff
@@ -688,6 +688,13 @@ static void pushCalibration2(double bgValue, String uuid, long offset) {
             GcmActivity.sendMessage(myIdentity(), "cal2", json);
         }
     }
+    public static void pushLibreBlock(String libreBlock) {
```
Comment (jamorham) [inline app/src/main/java/com/eveningoutpost/dexdrip/GcmActivity.java:691]: This should also have a `pratelimit` call to ensure that network traffic cannot increase much more than would be expected. So the rate limit should prevent the same or duplicate data being sent multiple times. I assume that this push should not occur more than once every approx 5 minutes?

File: app/src/main/java/com/eveningoutpost/dexdrip/Home.java [hunk 2]
```diff
@@ -3039,7 +3039,7 @@ public boolean onCreateOptionsMenu(Menu menu) {
         menu.findItem(R.id.showreminders).setVisible(Pref.getBoolean("plus_show_reminders", true) && !is_newbie);
 
         LibreBlock libreBlock = null;
-        if (DexCollectionType.hasLibre()) {
+        if (DexCollectionType.hasLibre() || get_follower()) {
```
Comment (jamorham) [inline app/src/main/java/com/eveningoutpost/dexdrip/Home.java:3042]: I think this additional use of get_follower needs to also be conditional on a preference switch to avoid this being activated for all followers. eg non-libre etc

File: app/src/main/java/com/eveningoutpost/dexdrip/Services/LibreWifiReader.java [hunk 3]
```diff
@@ -57,6 +58,9 @@
     private static final Gson gson = JoH.defaultGsonInstance();
 
     private final static long DEXCOM_PERIOD = 300000;
+    
+    // An object that is used to sync the readData
+    private static final Object readDataLock = new Object();
```
Comment (jamorham) [inline app/src/main/java/com/eveningoutpost/dexdrip/Services/LibreWifiReader.java:63]: Can you just do: `synchronized (LibreWifiReader.class)` to avoid creating an additional object?

File: app/src/main/java/com/eveningoutpost/dexdrip/Home.java (hunk 2 above)
Comment (tzachi-dar) [reply app/src/main/java/com/eveningoutpost/dexdrip/Home.java:3042]: Is it OK if I'll have this checked once per hour (with a rate limiter) or do you still prefer that I'll add a switch?

**Comment by tzachi-dar** [comment]
What about the extra preference to control the whole house coverage?
Should I add a preference that will only allow sending libreblocks?

File: app/src/main/java/com/eveningoutpost/dexdrip/GcmActivity.java (hunk 1 above)
Comment (tzachi-dar) [reply app/src/main/java/com/eveningoutpost/dexdrip/GcmActivity.java:793]: Yes, moved this to a more visible place.

File: app/src/main/java/com/eveningoutpost/dexdrip/Services/LibreWifiReader.java (hunk 3 above)
Comment (tzachi-dar) [reply app/src/main/java/com/eveningoutpost/dexdrip/Services/LibreWifiReader.java]: Done.

**Comment by tzachi-dar** [comment]
I have added my changes from the code review:

1) Send the sensor age to the follower.
2) Use sync on classes and not on separate objects.
3) Move the rate limiter on new libreblock objects to GcmActivity.java
4) Only notify for bad sensor if the sensor has not advanced, and there are no new readings.
5) Cache the results of the function hasLibreblock().
6) Fix a bug in the function checkLibreSensorChange.
7) Add a preference key to control if this host will be a whole house collector.

Still doing final tests, but so far things look well.

File: app/src/main/java/com/eveningoutpost/dexdrip/GcmActivity.java [hunk 4]
```diff
@@ -279,6 +279,10 @@ public synchronized static void syncBGReading(BgReading bgReading) {
             UserError.Log.wtf(TAG, "Cannot sync null bgreading - should never occur");
             return;
         }
+        if(Home.get_is_libre_whole_house_collector()) {
```
Comment (jamorham) [inline app/src/main/java/com/eveningoutpost/dexdrip/GcmActivity.java:282]: Why is this conditional here? What is it for?

File: app/src/main/java/com/eveningoutpost/dexdrip/Models/SensorSanity.java [hunk 5]
```diff
@@ -101,30 +102,109 @@ public static boolean checkLibreSensorChangeIfEnabled(final String sn) {
...
+                // This is the first time we see this sensor, update our uuid.
+                Log.i(TAG, "This is the first time we see this sensor uuid = " +  this_sensor.uuid);
+                PersistentStore.setString(PREF_LIBRE_SENSOR_UUID, this_sensor.uuid);
+                return false;
+            }
+        }
+    }
+
+    // This is a test for the CheckLibreSensorChange function.
+    public static void testCheckLibreSensorChange() {
```
Comment (jamorham) [inline app/src/main/java/com/eveningoutpost/dexdrip/Models/SensorSanity.java:159]: Should this be in its own unit test class?

File: app/src/main/java/com/eveningoutpost/dexdrip/NFCReaderX.java [hunk 6]
```diff
@@ -218,10 +219,32 @@ public static boolean HandleGoodReading(String tagId, byte[] data1, final long C
...
+        }
+        // Create the object to send
+        libreBlock = LibreBlock.create(tagId, CaptureDateTime, data1, 0, patchUid, patchInfo);
+        if(libreBlock == null) {
+            Log.e(TAG, "Error could not create libreBlock for libre-allhouse");
+            return;
+        }
+        final String json = libreBlock.toExtendedJson();
+        
+        GcmActivity.pushLibreBlock(json);
```
Comment (jamorham) [inline app/src/main/java/com/eveningoutpost/dexdrip/NFCReaderX.java:240]: This should be wrapped either within the function or the caller with both a preference switch that it only becomes active when libre all house is in use and a rate limit. It looks like you have neither from reading the PR?

File: app/src/main/java/com/eveningoutpost/dexdrip/GcmActivity.java [hunk 7]
```diff
@@ -688,8 +790,25 @@ static void pushCalibration2(double bgValue, String uuid, long offset) {
             GcmActivity.sendMessage(myIdentity(), "cal2", json);
         }
     }
+    public static void pushLibreBlock(String libreBlock) {
+        Log.i(TAG, "libreBlock called: " + libreBlock);
+        if (!Home.get_master()) {
+            return;
+        }
+        if (!JoH.ratelimit("libre-allhouse", 5)) {
```
Comment (jamorham) [inline app/src/main/java/com/eveningoutpost/dexdrip/GcmActivity.java:798]: Use `pratelimit` for critical rate limiting as `ratelimit` isn't safe on all platforms.

File: app/src/main/java/com/eveningoutpost/dexdrip/GcmListenerSvc.java [hunk 8]
```diff
@@ -552,7 +557,8 @@ public void run() {
                             UserError.Log.wtf(TAG, "Exception processing rsom timestamp");
                         }
                     }
-
+                } else if (action.equals("libreBlock")) {
+                    HandleLibreBlock(payload);
```
Comment (jamorham) [inline app/src/main/java/com/eveningoutpost/dexdrip/GcmListenerSvc.java:561]: Doesn't this call `NFCReaderX.HandleGoodReading` which then in turn calls `GcmActivity.pushLibreBlock` which causes the received data to be duplicated? Its very important that inbound message doesn't generate the same outbound message.

Comment (tzachi-dar) [reply app/src/main/java/com/eveningoutpost/dexdrip/GcmListenerSvc.java:561]: Messages should be unique by their timestamp.
In the function HandleLibreBlock we do the following:
1) We check if we have already seen this message (based on it's timestamp). If we did, we ignore it.
2) We save it in our DB.

3) On the function SendLibrereading() we check for a message with this timestamp again. if it is stored in our DB, we will not send it.
So, a message received from the network should not be sent again.

File: app/src/main/java/com/eveningoutpost/dexdrip/NFCReaderX.java (hunk 6 above)
Comment (tzachi-dar) [reply app/src/main/java/com/eveningoutpost/dexdrip/NFCReaderX.java:240]: Libreblocks have the data to needed to show libretrend, see sensor age, and device battery.
This information should pass to normal followers as well. So I did not limit this on a preference swithc.

In the function pushLibreBlock() we have the rate limiter.

File: app/src/main/java/com/eveningoutpost/dexdrip/GcmActivity.java (hunk 4 above)
Comment (tzachi-dar) [reply app/src/main/java/com/eveningoutpost/dexdrip/GcmActivity.java:282]: In order for an xDrip to use BT to get data it needs to have hardware_data_type == librebt
In order for him to send data it needs to be marked as master.

The question here is what will happen if the user will start a new sensor (and add calibrations).
If he will then the network will have two masters, each sending BG data (as well as other messages).

In order to only have one master on the network, I have limited this master to send messages of LibreBlock, alarm snoozes and 'desert mode' messages. This is why we need the get_is_libre_whole_house_collector condition.

An alternative to this is to write in documentation that the users will not start a new sensor, but then users will find it hard to tell if this xDrip is realy communicating with the miaomiao.

File: app/src/main/java/com/eveningoutpost/dexdrip/GcmActivity.java (hunk 7 above)
Comment (tzachi-dar) [reply app/src/main/java/com/eveningoutpost/dexdrip/GcmActivity.java]: Done.

File: app/src/main/java/com/eveningoutpost/dexdrip/Models/SensorSanity.java (hunk 5 above)
Comment (tzachi-dar) [reply app/src/main/java/com/eveningoutpost/dexdrip/Models/SensorSanity.java]: Yes it should.
I was not aware of this unit test, but moved it now.

**Comment by tzachi-dar** [comment]
I have addressed all comments.
Please take a look again.

File: app/src/main/java/com/eveningoutpost/dexdrip/Models/Bubble.java [hunk 9]
```diff
@@ -82,6 +82,11 @@ public static BridgeResponse decodeBubblePacket(byte[] buffer, int len) {
             patchUid = Arrays.copyOfRange(buffer, 2, 10);
             String SensorSn = LibreUtils.decodeSerialNumberKey(patchUid);
             PersistentStore.setString("LibreSN", SensorSn);
+            
+            if (SensorSanity.checkLibreSensorChangeIfEnabled(SensorSn)) {
+                Log.e(TAG, "Problem with Libre Serial Number - not processing");
+            }
```
Comment (jamorham) [inline app/src/main/java/com/eveningoutpost/dexdrip/Models/Bubble.java:88]: Should this do more than just log the problem? Was more intended?

File: app/build.gradle [hunk 10]
```diff
@@ -331,4 +331,12 @@ dependencies {
     annotationProcessor "org.projectlombok:lombok:1.18.10"
 }
 
+tasks.withType(Test) { 
+  testLogging {
+    exceptionFormat "full"
+    events "started", "skipped", "passed", "failed"
+    showStandardStreams true
+  }
+}
```
Comment (jamorham) [inline app/build.gradle:340]: Can you explain why this change is needed?

File: app/src/main/java/com/eveningoutpost/dexdrip/NFCReaderX.java [hunk 11]
```diff
@@ -218,10 +219,32 @@ public static boolean HandleGoodReading(String tagId, byte[] data1, final long C
         return HandleGoodReading(tagId, data1, CaptureDateTime, false, null, null);
     }
 
+
+    public static void SendLibrereading(final String tagId, byte[] data1, final long CaptureDateTime, byte []patchUid,  byte []patchInfo){
+        if(!Home.get_master()) {
+            return;
+        }
```
Comment (jamorham) [inline app/src/main/java/com/eveningoutpost/dexdrip/NFCReaderX.java:226]: Can we add a check for the librewholehouse preference switch here and return if it isn't set?

File: app/src/main/java/com/eveningoutpost/dexdrip/NFCReaderX.java [hunk 12]
```diff
@@ -282,6 +309,10 @@ protected void onPostExecute(Tag tag) {
                 if (succeeded) {
                     long now = JoH.tsl();
                     String SensorSn = LibreUtils.decodeSerialNumberKey(tag.getId());
+                    
+                    if (SensorSanity.checkLibreSensorChangeIfEnabled(SensorSn)) {
+                        Log.e(TAG, "Problem with Libre Serial Number - not processing");
```
Comment (jamorham) [inline app/src/main/java/com/eveningoutpost/dexdrip/NFCReaderX.java:314]: This logs that it isn't processing but does processing continue? should it?

File: app/build.gradle (hunk 10 above)
Comment (tzachi-dar) [reply app/build.gradle:340]: Without this change, when one is running the unit tests, he gets a summery at the end of what has ahppened.
With this change, one can also see the test results as they are running.

It will now look like (also with green color for passing tests.):

com.eveningoutpost.dexdrip.watch.thinjam.utils.BitmapToolsTest > unpackMonoBytesToRGB565Test STARTED

com.eveningoutpost.dexdrip.watch.thinjam.utils.BitmapToolsTest > unpackMonoBytesToRGB565Test PASSED

com.eveningoutpost.dexdrip.watch.thinjam.utils.BitmapToolsTest > packRGB565bytesToMonoTest STARTED

com.eveningoutpost.dexdrip.watch.thinjam.utils.BitmapToolsTest > packRGB565bytesToMonoTest PASSED

File: app/src/main/java/com/eveningoutpost/dexdrip/NFCReaderX.java (hunk 11 above)
Comment (tzachi-dar) [reply app/src/main/java/com/eveningoutpost/dexdrip/NFCReaderX.java:226]: This information is needed for all followers (not related to all house coverage).
It allows them to have per minute information (libre trend), sensor life and battery life.
Even Followers who will not use all house need this information.

File: app/src/main/java/com/eveningoutpost/dexdrip/NFCReaderX.java (hunk 12 above)
Comment (tzachi-dar) [reply app/src/main/java/com/eveningoutpost/dexdrip/NFCReaderX.java:314]: Sensor was stopped, so even if processing continues,  BG will not be created.
I can add a return in all this casses.

File: app/src/main/java/com/eveningoutpost/dexdrip/GcmActivity.java [hunk 13]
```diff
@@ -688,8 +790,25 @@ static void pushCalibration2(double bgValue, String uuid, long offset) {
             GcmActivity.sendMessage(myIdentity(), "cal2", json);
         }
     }
+    public static void pushLibreBlock(String libreBlock) {
+        Log.i(TAG, "libreBlock called: " + libreBlock);
+        if (!Home.get_master()) {
+            return;
+        }
+        if (!JoH.pratelimit("libre-allhouse", 5)) {
```
Comment (jamorham) [inline app/src/main/java/com/eveningoutpost/dexdrip/GcmActivity.java:798]: What is the expected frequency of these packets do you think?

Comment (tzachi-dar) [reply app/src/main/java/com/eveningoutpost/dexdrip/GcmActivity.java:798]: Well, if the user is continuously connected to the sensor, then probably every 5 minutes.
I guess that if he is near the device and then goes away, and then reconnects it reasonably should be every minute.
The goal was to block storms, so I figured out that 5 seconds should be enough.

**Comment by tzachi-dar** [comment]
This was replaced by:
https://github.com/NightscoutFoundation/xDrip/pull/1385
